apps/src/bin/quiche-server.rs

```diff
@@ -326,289 +341,383 @@ fn main() {
                     }
-                }
 
-                // Only bother with qlog if the user specified it.
-                #[cfg(feature = "qlog")]
-                {
-                    if let Some(dir) = std::env::var_os("QLOGDIR") {
-                        let id = format!("{:?}", &scid);
-                        let writer = make_qlog_writer(&dir, "server", &id);
-
-                        conn.set_qlog(
-                            std::boxed::Box::new(writer),
-                            "quiche-server qlog".to_string(),
-                            format!("{} id={}", "quiche-server qlog", id),
-                        );
+                    let scid = quiche::ConnectionId::from_vec(scid.to_vec());
+
+                    trace!("New connection: dcid={:?} scid={:?}", hdr.dcid, scid);
+
+                    let client_id = next_client_id;
+
+                    // Set the preferred address transport parameter
+                    let rng = SystemRandom::new();
+                    let (preferred_address_scid, preferred_address_reset_token) =
+                        generate_cid_and_reset_token(&rng);
+
+                    match preferred_address {
+                        SocketAddr::V4(preferred_addr_v4) => config
+                            .set_preferred_address(
+                                Some(preferred_addr_v4),
+                                None,
+                                preferred_address_scid.clone(),
+                                preferred_address_reset_token,
+                            )
+                            .unwrap(),
+                        SocketAddr::V6(preferred_addr_v6) => config
+                            .set_preferred_address(
+                                None,
+                                Some(preferred_addr_v6),
+                                preferred_address_scid.clone(),
+                                preferred_address_reset_token,
+                            )
+                            .unwrap(),
                     }
-                }
 
-                let client_id = next_client_id;
-
-                let client = Client {
-                    conn,
-                    http_conn: None,
-                    client_id,
-                    partial_requests: HashMap::new(),
-                    partial_responses: HashMap::new(),
-                    app_proto_selected: false,
-                    max_datagram_size,
-                    loss_rate: 0.0,
-                    max_send_burst: MAX_BUF_SIZE,
-                };
+                    // Track the preferred address connection ID with the client
+                    // it was given to
+                    clients_ids.insert(preferred_address_scid, client_id);
+
+                    // Create the new connection!
+                    #[allow(unused_mut)]
+                    // Automatically provisions SCID for preferred address.
+                    let mut conn = quiche::accept(
```

Comment:
Is this a hangover? if we don't need the `mut` then lets not declare it as such

Reply:
`conn.set_keylog(Box::new(keylog));` needs a mutable ref